Provide automated way of updating repo project boards.

brahmaroutu opened the pull request:
This is the PR to add project automation capabilities.This code is picked up from https://github.com/kubernetes/test-infra/pull/10401 and would like to add this PR to enable project automation and manage conformance related work. 

https://github.com/kubernetes/test-infra/pull/10401 cannot be merged in its state as I did not get response form the author.
area/conformance 
sig/testing

taragu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,204 @@
+	return nil
+}
+
+func getMatchingColumnIDs(gc githubClient, projectManager plugins.ProjectManager, pe github.PullRequestEvent) ([]int, error) {
```
- Looks like we are only using `projectManager.OrgRepos` in this method, so we can pass that in instead of the entire `plugins.ProjectManager`
- We can call `gc.GetIssueLabels(...)` in `handlePR(...)` instead of calling `gc.GetIssueLabels(...)` in the nested for loops.

brahmaroutu replied:
@taragu Not true, projectManager has a set of Managed Projects and a set of ManagedColumns embedded, we need to pass the entire structure to validate if the PR meets the criteria.

taragu replied:
`projectManager has a set of Managed Projects and a set of ManagedColumns embedded` --> @brahmaroutu  that's right. Managed repos, projects, columns are all under `projectManager.OrgRepos`. So can we pass `projectManager.OrgRepos` into this function instead of `projectManager`?

brahmaroutu replied:
Done

brahmaroutu replied:
I am not sure if there is any advantage to passing in  map[string]ManagedOrgRepo vs ProjectManager object?

taragu replied:
@brahmaroutu I suggested this for better readability. If we pass in `map[string]ManagedOrgRepo` then readers immediately know from the function signature that this is the piece of data being used, versus if we pass in `plugins.ProjectManager`, readers will need to read the content of this function to see that only `projectManager.OrgRepos` is used. This is minor and you can leave this as is. It's completely up to you if you would like to make this change.

taragu commented on prow/plugins/projectmanager/projectmanager_test.go:
```diff
@@ -0,0 +1,284 @@
+	"k8s.io/test-infra/prow/plugins"
+)
+
+type fakeGitHubClient struct {
```
Can we use [the existing fakegithub methods](https://github.com/kubernetes/test-infra/blob/master/prow/github/fakegithub/fakegithub.go) instead of creating new ones here?

brahmaroutu replied:
Done

brahmaroutu commented:
@stevekuznetsov @cjwagner Can you please review
/assign cjwagner

brahmaroutu commented:
This PR is tested with the following configuration
project_manager:
  org/repos: 
    ii/apisnoop:
      projects:
        test_project:
          columns:
          - id: 0
            name: triage
            state: open
            org: ii
            labels: 
            - area/conformance
          - id: 0
            name: need-attention
            state: open
            org: ii
            labels: 
            - area/apisnoop
        test_project2:
          columns:
          - id: 0
            name: triage
            state: open
            org: ii
            labels:
            - area/conformance2
          - id: 0
            name: need-attention
            state: open
            org: ii
            labels: 
            - area/apisnoop2

When a PR is labeled '/area conformance' it automatically created card under the test_project. 
Please note that id:0 is set in order to obtain the real id of the column using the column name.

brahmaroutu commented:
@timothysc Please note that I have basic functionality working with PR using above mentioned configuration.

stevekuznetsov commented on prow/plugins/config.go:
```diff
@@ -437,6 +438,34 @@ type ProjectRepoConfig struct {
 	ProjectColumnMap map[string]string `json:"repo_default_column_map,omitempty"`
 }
 
+// ProjectManager represents the config for the ProjectManager plugin, holding top
```
Godoc seems stale already

brahmaroutu replied:
Sorry?, The Good explain what a project manager does. It is a automated way to enable PR's to be added to a project if the project is configured.

stevekuznetsov replied:
The struct already does not hold a list of projects as the GoDoc claims

brahmaroutu replied:
modified the statement, please see.

stevekuznetsov commented on prow/plugins/config.go:
```diff
@@ -437,6 +438,34 @@ type ProjectRepoConfig struct {
+// ProjectManager represents the config for the ProjectManager plugin, holding top
+// level config options and a list of Projects
+type ProjectManager struct {
+	OrgRepos map[string]ManagedOrgRepo `json:"org/repos,omitempty"`
```
I don't think we want the key to be `"org/repos"` -- maybe have:

```go
type Configuration struct {
    // ...
    ProjectManagementConfig map[string]MangedOrgRepo `json:"project_management,omitempty"`
}
```

?

brahmaroutu replied:
how about "org_repo"?

stevekuznetsov replied:
We want to have the key be the _name_ of the thing happening, not what it is. There are many configuration options that are org/repo format.

brahmaroutu replied:
cool, changed it as suggested.

stevekuznetsov commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,204 @@
+	pluginName = "project-manager"
+)
+
+// TODO Create a new handler for issues, look in hook/server.go
```
outstanding? or remove the TODO

brahmaroutu replied:
it is outstanding, I will work on it after this PR.

stevekuznetsov commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,204 @@
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The project-manager plugin automatically adds Pull Requests to specified GitHub Project Columns if they match given criteria.",
+		Config: func(config *plugins.Configuration) map[string]string {
```
Please provide a human-readable configuration summary here

brahmaroutu replied:
done.

stevekuznetsov commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,204 @@
+					}
+				}
+				columnID := managedColumn.ID
+				// Currently this assumes columnID having a value if 0 means it is unset
```
You can use a `*int` here and check `nil` instead of special-casing `0` if you'd like

brahmaroutu replied:
I need to have a default value to look for. It is either 0 or -1 which can never be a valid column id, *int will make me have configuration omitting the column, unless there is a good need for all that code change?

stevekuznetsov replied:
Why do you need to have a default value? You are saying you need to check to see if the value is set or not. Using a pointer and checking if it is `nil` is the idiomatic way to check if a value has or has not been set.

brahmaroutu replied:
changed code to use int*

stevekuznetsov commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+// hasLabels checks if all labels are in the github.label set "issueLabels"
+func hasLabels(labels []string, issueLabels []github.Label) bool {
```
Lets add this to the utils file

brahmaroutu replied:
Done

stevekuznetsov commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+func handlePR(gc githubClient, projectManager plugins.ProjectManager, log *logrus.Entry, pe github.PullRequestEvent) error {
+	// Only respond to label add or issue/PR open events
```
Wait, right now this is not checking for any comment command -- are we trying to add every single opened PR and issue to a board?

brahmaroutu replied:
If a project is configured and a label is added to the PR, we add the PR to that project under the column mentioned or in a default column on the project.

stevekuznetsov commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,204 @@
+						// If labels are not yet fetched then get them as they are now required
+						// GetIssueLabels works for PRs as they are considered issues in the API
+						var err error
+						labels, err = gc.GetIssueLabels(pe.Repo.Owner.Login, pe.Repo.Name, pe.Number)
```
IIRC we get the labels on the event? So we do not need to do this fetch?

brahmaroutu replied:
Yes, if we just react to PullRequestEvents with Action 'labeled' which will be triggered for each of the labels. We only have to perform necessary action for that label. Currently code processes 'Open' and other PR Actions and we should ignore those. WDYT

brahmaroutu commented:
@stevekuznetsov, @cjwagner  please review.

brahmaroutu commented:
@stevekuznetsov @cjwagner I would like to have this functionality ASAP, conformance project board want to experiment using this approach to automatically add the PR/Issue to the project. I also have fix for issue once this original PR is merged.
Please review and let me know if I have addresses all the concerns.

stevekuznetsov commented:
I don't think I'll have time to do another pass this week, sorry -- @cjwagner could you please take a look?

cjwagner commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,200 @@
+	case 1:
+		projects, err = gc.GetOrgProjects(orgRepoParts[0])
+	default:
+		return nil, fmt.Errorf("could not determine org or org/repo from %s", orgRepoName)
```
Ideally we would include this validation in the plugin config parsing logic so that this fails when config is loaded instead of when we try to use the value.

brahmaroutu replied:
I think the yaml cannot be valid if org/repo is not specified,  case 1 or case 2 are only possibilities. I would also like to know if there is a callback to validate the configuration within plugin.

cjwagner commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,200 @@
+	// For each ManagedColumn that matches this PR, add this PR to that Project Column
+	for _, matchedColumnID := range matchedColumnIDs {
+		err = addPRToColumn(gc, matchedColumnID, pe)
+		log.WithError(err).Println("Failed to add PR to project")
```
Please use `Error` instead of `Println` and also include relevant log fields using `WithFields` e.g. `pr` and `project-column-id`.

brahmaroutu replied:
Done

cjwagner commented on prow/plugins/projectmanager/projectmanager_test.go:
```diff
@@ -0,0 +1,257 @@
+			},
+		},
+		{
+			name: "add pull request to project column with only columnID",
```
This test case has more than just `columnID` specified.
The `expectedColumnCards` is also identical to the previous test case's so this test case doesn't actually appear to be validating the behavior. (I would expect a different expected column ID to be used for this test case so that we can distinguish the behavior in these two cases.)

Additionally, the value of `columnID` is 0, but the test case is still passing even though the `expectedColumnCards` indicates that a card is added to column 1. This looks like a bug that the test failed to detect. Please confirm if this is the case and update the code and test if it is.

brahmaroutu replied:
Modified the tests, tested with column ID's for multiple projects

cjwagner commented on prow/plugins/projectmanager/projectmanager_test.go:
```diff
@@ -0,0 +1,257 @@
+			},
+			expectedColumnCards: map[int][]github.ProjectCard{},
+		},
+	}
```
Please add test cases for
- config for multiple repos in the same org (show that only the relevant config applies)
- config at both the org and repo level (show that both levels apply)
- a PR that should add multiple column cards

brahmaroutu replied:
Done

cjwagner commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,200 @@
+	// Only respond to label add or issue/PR open events
+	if pe.Action != github.PullRequestActionOpened &&
+		pe.Action != github.PullRequestActionReopened &&
+		pe.Action != github.PullRequestActionLabeled {
```
What happens if the PR was already added to the project and a label is added or the PR is reopened? It seems like this will create a duplicate column card (or fail trying if that is disallowed). This is particularly problematic for the "label added" case which is very common.

Similarly, what happens if a PR stops meeting the criteria? (state change, label change, etc.) Currently this code doesn't attempt to handle that case so project cards are never removed.

brahmaroutu replied:
Currently we cannot add project card to the project if it already exist. Duplication is checked by Git API.  I modifies the fakegithub client to behave similarly
If a label is removed we need to handle the case to remove the project card, this is stated in the TODO items at the top of the module.

cjwagner replied:
> Duplication is checked by Git API

IIUC every time a label is added we'll try to add all the column cards. For each of the column cards we waste an API token trying to add the card again when it already exists. Since this can happen for every label event this could be a significant waste, right? Can we cost efficiently find all the column cards for the PR and deduplicate ourselves or would that require the same number of API requests?

brahmaroutu replied:
I think I understand the concern,  git project allows users to add note cards on issue/Prs from any org and repo. This allows projects to track external dependencies.
I wish the PR/Issue api is enhanced to provide info on all projects it belongs to (like how the UI shows in the right side column)
Currently, I believe, the cost of failing with 'Card already exists' is cheaper than maintaining a preloaded cache and maintaining refresh.  This can be improved later too.

brahmaroutu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,200 @@
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The project-manager plugin automatically adds Pull Requests to specified GitHub Project Columns if they match given criteria.",
+		Config: map[string]string{"project_manager": `orgsRepos: 
```
Done

brahmaroutu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,200 @@
+	// Create project card and add this PR
+	projectCard := github.ProjectCard{}
+	projectCard.ContentType = "PullRequest"
+	projectCard.ContentID = pe.PullRequest.ID
```
Done

brahmaroutu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,200 @@
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The project-manager plugin automatically adds Pull Requests to specified GitHub Project Columns if they match given criteria.",
```
Done

brahmaroutu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,200 @@
+                                   columns:
+                                     - name: triage
+                                     state: open
+                                     org: kubernetes
```
Sometimes Projects in K/K might contain cards with issues/PRs from community, website and other repos. Example "Add KEP for Behavior-driven Conformance Testing #959" belongs to Kubernetes/enhancements but this card is under conformance project.

cjwagner replied:
I don't understand the example you provided. How is it associated with `k/enhancements` and what conformance project are you referring to?
Do we need to allow repo to be specified too?
Is it possible to have a card reference an issue/PR in a different org? (not just a different repo)

brahmaroutu replied:
This code is based on original PR #10401, I believe with the author that we should consider projects that may contain cards from other orgs. May be not very realistic for k/k projects. Please let me know if you think that we should restrict to make this to work with projects inside k/k only.

brahmaroutu commented:
@cjwagner sorry for the delay due to travel, KubeCon etc. Please review and see if I addressed all the concerns.

brahmaroutu commented:
@cjwagner @stevekuznetsov Please review. I will have another PR to handle Issues and for Label removal logic.

cjwagner commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,208 @@
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The project-manager plugin automatically adds Pull Requests to specified GitHub Project Columns, if the label on the PR matches with configured project and the column.",
+		Config: map[string]string{
```
This isn't how `Config` is meant to be used. Help providers are functions that take the current config and the list of enabled repos so that they can return dynamically generated help information specific to the current configuration.
`Config` needs to describe the current configuration, not provide a hard coded configuration example.
Here is an example from the `approve` plugin: https://github.com/kubernetes/test-infra/blob/be2fa898afec153f9d46f17c95a3f595d97dc29e/prow/plugins/approve/approve.go#L114-L134

brahmaroutu replied:
I am a bit confused and not sure how to test this locally, but I changed code to create a readable structure from the 'Config' provided.

cjwagner commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,208 @@
+	}
+	_, err := gc.CreateProjectCard(columnID, projectCard)
+	if err != nil {
+		return err
```
This will return errors when we add a duplicate card, right? If that is expected behavior we need to distinguish the duplicate card error and avoid returning an error in that case.

brahmaroutu replied:
Done

brahmaroutu commented:
@cjwagner Please review and see if your concerns are addressed.

brahmaroutu commented:
@cjwagner @stevekuznetsov  Can you please review. I would like to get this functionality in ASAP.

cjwagner commented on prow/plugins/config.go:
```diff
@@ -448,6 +449,35 @@ type ProjectRepoConfig struct {
 
+// ProjectManager represents the config for the ProjectManager plugin, holding top
+// level config options, configuration is a hierarchial structure with top level element
+// being org/repo with the list of projects as its children
```
nit: "org" or "org/repo"

brahmaroutu replied:
Fixed

cjwagner commented on prow/plugins/config.go:
```diff
@@ -448,6 +449,35 @@ type ProjectRepoConfig struct {
+	Name   string   `json:"name,omitempty"`
+	State  string   `json:"state,omitempty"`
+	Labels []string `json:"labels,omitempty"`
+	Org    string   `json:"org,omitempty"`
```
I feel like this config might be backwards? This struct both describes the column (Name, ID) and the conditions that the PR must match (Org) while the column's own Org is actually specified at a higher config level. This distinction is not clear from the GoDocs, please improve them to make this clear.

I also still don't understand why `Org` is specified at this level, but `Repo` is not.

cjwagner commented on prow/plugins/config.go:
```diff
@@ -448,6 +449,35 @@ type ProjectRepoConfig struct {
+type ManagedColumn struct {
+	ID     *int     `json:"id,omitempty"`
+	Name   string   `json:"name,omitempty"`
+	State  string   `json:"state,omitempty"`
```
Config loading validation checks we should have:
- At least one of `id` and `name` must be specified.
- `state` must be empty or a valid state string. (Please mention the accepted values in a comment here as well.)
- The keys of `orgsRepos` should be valid "org" or "org/repo" strings.

cjwagner commented on prow/github/fakegithub/fakegithub.go:
```diff
@@ -507,6 +509,16 @@ func (f *FakeClient) CreateProjectCard(columnID int, projectCard github.ProjectC
+			// Create project will generate an error when the card already exist in the project
+			if _, exists := columnIDMap[columnID]; exists {
+				card, err := f.GetColumnProjectCard(id, projectCard.ContentURL)
+				if err == nil && card != nil && *card == projectCard {
```
I think there is still a collision even if `*card != projectCard`.

brahmaroutu replied:
There can be only one project card per issue/PR and since we are looking through all columns in a project to see if there exists a projectCard  we should be safe once this check passes. If you are worried about collision due to concurrency, we will have a failure from the CreateCard api itself.

cjwagner commented on prow/github/fakegithub/fakegithub.go:
```diff
@@ -507,6 +509,16 @@ func (f *FakeClient) CreateProjectCard(columnID int, projectCard github.ProjectC
+		for id := range columnIDMap {
+			// Make sure that we behave same as github API
+			// Create project will generate an error when the card already exist in the project
+			if _, exists := columnIDMap[columnID]; exists {
```
nit: It is strange to put this conditional inside the loop given that it will evaluate the same for every loop iteration. Please move outside the loop.

brahmaroutu replied:
fixedtrue, fixed.

cjwagner commented on prow/plugins/projectmanager/projectmanager_test.go:
```diff
@@ -0,0 +1,678 @@
+				},
+			},
+			expectedColumnCards: map[int][]github.ProjectCard{},
+			expectedError:       fmt.Errorf("could not find column %s in project %s", "testColumn", "testProject"),
```
It is unclear to me why this testcase called "don't add pull request with incorrect labels" would result in an error. We don't want to emit errors on perfectly valid PRs that don't need to be touched by this plugin.

brahmaroutu replied:
Changed the test name, I changed code to not return error but ignore adding the PR to the project if that makes sense. It is possible to drop columns from projects after we configure and there is possibility that we cannot add the project card, we should ignore and log the details.

cjwagner commented on prow/plugins/projectmanager/projectmanager_test.go:
```diff
@@ -0,0 +1,678 @@
+					return
+				}
+				t.Fatalf("handlePR error: %v", err)
+			}
```
Please fail the test if we expect an error and don't see one.

brahmaroutu replied:
fixed.

cjwagner commented on prow/plugins/projectmanager/projectmanager_test.go:
```diff
@@ -0,0 +1,678 @@
+			},
+		},
+		{
+			name: "add pull request using column name to multiple columns in a project should fail",
```
Why is there no `expectedError` field specified for this test case? This makes me worry that the tests are not actually validating the plugin properly.

brahmaroutu replied:
The fix currently is to not throw errors if we fail to add a project card, instead we log the issue why we failed to add and proceed.

timothysc commented:
@brahmaroutu - can you quickturn on this and we can finish it.  This has taken O(months).

brahmaroutu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,201 @@
+	for orgRepoName, managedOrgRepo := range config.ProjectManager.OrgRepos {
+		configString = fmt.Sprintf("%s %s: { projects: {", configString, orgRepoName)
+		for projectName, managedProject := range managedOrgRepo.Projects {
+			configString := fmt.Sprintf("%s %s: { columns: [", configString, projectName)
```
Please check the new iteration, hopefully addressed these concerns

guineveresaenger commented:
I am eager to try using this plugin for sig-release!

cjwagner commented:
Let me know when everything has been addressed and I'll do another review.

brahmaroutu commented:
@cjwagner Can you please review.

cjwagner commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,310 @@
+	if err != nil {
+		return err
+	}
+	return nil
```
nit:
```go
if err != nil {
    return err
}
return nil
```
==
```go
return err
```

brahmaroutu replied:
done

brahmaroutu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,310 @@
+
+				for _, card := range cards {
+					if card.ContentURL == issueURL {
+						return nil, errors.New(fmt.Sprintf(issueAlreadyInProject, issueURL, projectName))
```
getColumnID either return a valid columnID or an error explaining why it could not return a valid columnID. Do you expect caller to check for nil on columnID instead of an error?

brahmaroutu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,310 @@
+	var matchedColumnIDs []int
+	// Don't use GetIssueLabels unless it's required and keep track of whether the labels have been fetched to avoid unnecessary API usage.
+	var labels []github.Label
+	// If labels are not yet fetched then get them as they are now required
```
fixed

brahmaroutu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,310 @@
+				configErr = mustHaveColumns
+				continue
+			}
+			repoConfig = fmt.Sprintf("projects: {%s %s: { columns: [", repoConfig, projectName)
```
Redid the code, see if that works.

brahmaroutu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,310 @@
+	if len(projectConfig.OrgRepos) == 0 {
+		pluginHelp := &pluginhelp.PluginHelp{
+			Description: "The project-manager plugin automatically adds Pull Requests to specified GitHub Project Columns, if the label on the PR matches with configured project and the column.",
+			Config:      map[string]string{"Config": sampleConfig},
```
fixed.

brahmaroutu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,310 @@
+	for orgRepoName, managedOrgRepo := range config.ProjectManager.OrgRepos {
+		orgRepoParts := strings.Split(orgRepoName, "/")
+		if len(orgRepoParts) != 2 && len(orgRepoParts) != 1 {
+			configString[orgRepoName] = fmt.Sprintf(invalidOrgRepo, orgRepoName)
```
I moved validation here. Please review.

brahmaroutu commented on prow/github/helpers.go:
```diff
@@ -33,6 +33,16 @@ func HasLabel(label string, issueLabels []Label) bool {
+func HasLabels(labels []string, issueLabels []Label) bool {
+	for _, label := range labels {
+		if HasLabel(label, issueLabels) {
+			return true
```
fixed, added a test case with more than one label to match.

timothysc commented:
@cjwagner - This PR has been open since April and is needed for enable the community.  I'd recommend to merge and where needed to fixup afterwards.  

/assign @spiffxp

brahmaroutu commented on prow/plugins/config.go:
```diff
@@ -918,6 +952,45 @@ func validateRequireMatchingLabel(rs []RequireMatchingLabel) error {
+
+	projectConfig := pm
+	if len(projectConfig.OrgRepos) == 0 {
+		return fmt.Errorf("The project-manager plugin configuration is empty")
```
Fixed

brahmaroutu commented on prow/plugins/config.go:
```diff
@@ -918,6 +952,45 @@ func validateRequireMatchingLabel(rs []RequireMatchingLabel) error {
+				s_set := sets.NewString(managedColumn.Labels...)
+				for _, labels := range labelSets {
+					if s_set.Equal(sets.NewString(labels...)) {
+						return fmt.Errorf("Org/repo: %s, project %s, column %s has same labels configured as another column", orgRepoName, projectName, managedColumn.Name)
```
Fixed

brahmaroutu commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,244 @@
+	for orgRepoName, managedOrgRepo := range config.ProjectManager.OrgRepos {
+		for projectName, managedProject := range managedOrgRepo.Projects {
+			for _, managedColumn := range managedProject.Columns {
+				repoDescr = fmt.Sprintf("%s\nPR/Issues org: %s, with matching labels: %s will be added to the project: %s\n", repoDescr, managedColumn.Org, managedColumn.Labels, projectName)
```
Fixed handler, added State into the printout.

brahmaroutu commented on prow/plugins/config.go:
```diff
@@ -918,6 +952,45 @@ func validateRequireMatchingLabel(rs []RequireMatchingLabel) error {
+					return fmt.Errorf("Org/repo: %s, project %s, column %s, has no labels configured", orgRepoName, projectName, managedColumn.Name)
+				}
+				if len(managedColumn.Org) == 0 {
+					return fmt.Errorf("Org/repo: %s, project %s, column %s, has no org configured", orgRepoName, projectName, managedColumn.Name)
```
I have changed the code in getMatchingColumnIDs to ignore the column if Org is empty. The label matching logic with return false if the labels are empty on the column.
